create-diff-object: check for out-of-range relocations

sm00th opened the pull request:
Check that none of the relocations are out-of-range of their
corresponding sections before writing the output elf.

Fixes: #618

Signed-off-by: Artem Savkov <asavkov@redhat.com>

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -1923,6 +1923,28 @@ static void kpatch_regenerate_special_section(struct kpatch_elf *kelf,
+		list_for_each_entry_safe(rela, safe, &sec->relas, list) {
+			if (rela->sym->sec) {
+				sdata = rela->sym->sec->data;
+				if (rela->addend < sdata->d_off ||
```
On x86, rela->addend is often negative in relation to the symbol, e.g.:
```
Relocation section [20] '.rela.text.semctl_stat' for section [19] '.text.semctl_stat' at offset 0x720 contains 6 entries:
  Offset              Type            Value               Addend Name
  0x0000000000000001  X86_64_NONE     000000000000000000      -4 __fentry__
  0x000000000000002d  X86_64_PC32     000000000000000000      -4 ipc_obtain_object_check
  0x000000000000004f  X86_64_PC32     000000000000000000      -4 ipcperms
  0x0000000000000062  X86_64_PC32     000000000000000000      -4 security_sem_semctl
  0x0000000000000074  X86_64_PC32     000000000000000000      -4 ipc_obtain_object_idr
  0x0000000000000090  X86_64_PC32     000000000000000000      -4 kernel_to_ipc64_perm
```
That's because on x86, the relative address is adjusted by the length of the instruction which uses it.

So for X86_64_PC32 relocations, the check may need to take into account the length of the instruction which uses the rela's address.  Though that's probably overkill.  Instead I would recommend just removing the `<` check, since I don't think we don't change the addend anyway and I don't see how it could go beyond the beginning of the section.

The `>` check is much more important IMO.

jpoimboe commented on kpatch-build/create-diff-object.c:
```diff
@@ -1923,6 +1923,28 @@ static void kpatch_regenerate_special_section(struct kpatch_elf *kelf,
+			if (rela->sym->sec) {
+				sdata = rela->sym->sec->data;
+				if (rela->addend < sdata->d_off ||
+						rela->addend > (sdata->d_off + sdata->d_size)) {
```
Realistically, I believe there's always only *one* data descriptor per section and `d_off` will always be zero, so it should be fine to do `if (rela->addend >= sdata->d_size)`.

sm00th replied:
Should this be >= or >? I've been testing with the patch from #615 on top of current Linus' tree and am getting a relocation that would fail >= check. The resulting module does load alright though:

```
  [12] .text.compat_SyS_getdents PROGBITS        0000000000000000 0002d0 00012f 00  AX  0   0 16
                                                                                   
Relocation section '.rela.debug_ranges' at offset 0x39850 contains 34 entries:  
    Offset             Info             Type               Symbol's Value  Symbol's Name + Addend
...                                                                                
00000000000004d8  0000000a00000001 R_X86_64_64            0000000000000000 .text.compat_SyS_getdents + 12f 
...                                                                                

```

jpoimboe replied:
Ah, that's one of the DWARF debug sections.  I think '>' is probably fine then.

sm00th commented:
v2 changes:
  - list_for_each_entry_safe -> list_for_each_entry since the list is never modified
  - removed lower bound check
  - removed data offset from checks since it is expected to always be 0
  - added addend to the error message
